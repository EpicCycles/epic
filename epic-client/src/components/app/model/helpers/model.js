import {
    findObjectWithKey,
    generateRandomCode,
    isItAnObject,
    removeKey,
    updateObject,
    updateObjectInArray
} from "../../../../helpers/utils";
import {NEW_ELEMENT_ID} from "../../../../helpers/constants";

export const eliminateReadOnlyFields = (fieldList) => {
    return fieldList.filter(field => (! field.readOnly));
};

export const checkForChanges = (fieldList, existingObject, newValues) => {
    return eliminateReadOnlyFields(fieldList).some(field => {
        return existingObject[field.fieldName] !== newValues[field.fieldName];
    })
};

export const addFieldToState = (initialState, fieldList, fieldName, input) => {
    let finalState = initialState;
    fieldList.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            finalState = applyFieldValueToModel(initialState, field, input);
            return true;
        }
        return false;
    });
    return finalState;
};


export const getModelKey = modelInstance => {
    if (!modelInstance) return NEW_ELEMENT_ID;
    if (modelInstance.id) return modelInstance.id;
    if (modelInstance.dummyKey) return modelInstance.dummyKey;
    return NEW_ELEMENT_ID;
};

export const isModelValid = (modelInstance) => {
    return !(modelInstance.error || isItAnObject(modelInstance.error_detail));
};

export const applyFieldValueToModel = (modelInstance, field, value) => {
    let updatedModelInstance = updateObject(modelInstance);
    updatedModelInstance[field.fieldName] = value;
    updatedModelInstance.changed = true;
    if (!updatedModelInstance.error_detail) updatedModelInstance.error_detail = {};
    if (field.required && !value) {
        updatedModelInstance.error_detail[field.fieldName] = field.error;
        return updatedModelInstance;
    } else if (field.validator) {
        const error = field.validator(value, modelInstance);
        if (error) {
            updatedModelInstance.error_detail[field.fieldName] = error;
            return updatedModelInstance;
        }
    }
    updatedModelInstance.error_detail = removeKey(updatedModelInstance.error_detail, field.fieldName);
    return updatedModelInstance;
};

export const getAttribute = (modelFields, fieldName) => {
    let attribute;
    modelFields.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            attribute = field.fieldName;
            return true;
        }
        return false;
    });
    return attribute;
};

export const getField = (modelFields, fieldName) => {
    let modelField;
    modelFields.some(field => {
        if (fieldName.startsWith(field.fieldName)) {
            modelField = field;
            return true;
        }
        return false;
    });
    return modelField;
};

export const updateModelChanges = (modelInstance, attribute, fieldValue) => {
    let changes = updateObject(modelInstance.changes);
    if (fieldValue && (modelInstance[attribute] !== fieldValue)) {
        changes[attribute] = fieldValue;
    } else {
        changes = removeKey(changes, attribute);
    }
    return updateObject(modelInstance, { changes })
};
export const updateModelWithChanges = (modelInstance, modelFields, fieldName, fieldValue) => {
    const attribute = getAttribute(modelFields, fieldName);
    if (attribute) {
        return updateModelChanges(modelInstance, attribute, fieldValue)
    } else {
        console.error("attribute not found for", modelFields, fieldName)
    }
    return modelInstance;
};
export const updateModelArrayWithChanges = (modelArray, modelFields, fieldName, fieldValue, componentKey) => {
    const attribute = getAttribute(modelFields, fieldName);
    if (attribute) {
        let modelInstance = findObjectWithKey(modelArray, componentKey);
        let currentChanges = modelInstance.changes || {};
        if (fieldValue && (modelInstance[attribute] !== fieldValue)) {
            currentChanges[attribute] = fieldValue;
        } else {
            currentChanges = removeKey(currentChanges, attribute);
        }
        modelInstance.changes = currentChanges;
        return updateObjectInArray(modelArray, modelInstance, componentKey)
    } else {
        console.error("attribute not found for", modelFields, fieldName)
    }
    return modelArray;
};
export const updateModel = (model, modelFields, fieldName, fieldValue, componentKey) => {
    const modelField = getField(modelFields, fieldName);
    if (modelField) return applyFieldValueToModel(model, modelField, fieldValue);
    return model;
};

export const displayModelErrorSummary = (model, modelFields) => {
    let displayErrors = model.error;
    const fieldErrors = isItAnObject(model.error_detail) ? model.error_detail : {};
    for (var property in fieldErrors) {
        const field = getField(modelFields, property);
        if (!displayErrors) {
            displayErrors = "";
        } else {
            displayErrors += '<br>';
        }
        displayErrors += field.header + ": " + fieldErrors[property].join(" ");
    }

    return displayErrors;
};

export const getNameForValue = (value, sourceArray) => {
    if (!value) return undefined;
    let name = value;

    sourceArray.some(pair => {
        if (pair.value === value) {
            name = pair.name;
            return true;
        }
        return false;
    });
    return name;
};

export const hasErrors = (model) => {
    if (model.error && model.error.length > 0) return true;
    if (isItAnObject(model.error_detail)) return true;
    return false;
};

export const addErrorDetail = (error_detail = {}, field, error) => {
      let updated_error_detail = updateObject(error_detail);
  if (!Array.isArray(updated_error_detail[field])) {
         updated_error_detail[field] = [];
  }
  updated_error_detail[field].push(error);
  return updated_error_detail;
};

export const createNewModelInstance = () => {
    return {dummyKey: generateRandomCode()};
};
export const matchesModel = (persistedModel, modelFields, modelToCheck) => {
     return eliminateReadOnlyFields(modelFields).every(field => {
        return persistedModel[field.fieldName] === modelToCheck[field.fieldName];
    })
};