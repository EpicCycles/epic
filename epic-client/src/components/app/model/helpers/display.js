export const fixedHeaderClassname = (lockColumn) => {
    if (lockColumn) return "grid-header--fixed-left";
    return "";
};
export const fixedDetailsClassname = (lockColumn) => {
    if (lockColumn) return "grid-item--fixed-left";
    return "";
};
export const gridHeaderClass = (baseClassName = "", fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && (fieldIndex === 0);
    return `${baseClassName} grid-item--header ${fixedHeaderClassname(shouldLock)}`;
};
export const gridItemClass = (baseClassName = "", fieldIndex, firstColumnLocked) => {
  const shouldLock = firstColumnLocked && (fieldIndex === 0);
    return `${baseClassName} grid-item ${fixedDetailsClassname(shouldLock)}`;
};