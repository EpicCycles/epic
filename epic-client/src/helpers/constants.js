export const attributeOptionTypes = [
    {value: '1', name: 'Text', isDefault: true},
    {value: '2', name: 'Numeric'},
    {value: '3', name: 'Single - Radio'},
    {value: '4', name: 'Single - Dropdown'},
    {value: '5', name: 'Multiple - Radio'},
    {value: '6', name: 'Multiple - Dropdown'},
];

export const NEW_ELEMENT_ID = "new";

export const menuStructure = [
    {
        sectionPos: 1,
        sectionContents: [
            {
                groupHeader: "Customer",
                groupPos: 10,
                groupLinks: [
                    {displayText: "Find Customer", linkRoute: "/customer-search", linkNumber: 101},
                    {displayText: "New Customer", linkRoute: "/customer", linkNumber: 102},
                ]
            },
        ]
    },
    {
        sectionPos: 2,
        sectionContents: [
            {
                groupHeader: "Bikes",
                groupPos: 20,
                groupLinks: [
                    {displayText: "Bike Upload", linkRoute: "/bike-upload", linkNumber: 201},
                ]
            },
            {
                groupHeader: "Core Data",
                groupPos: 21,
                groupLinks: [
                    {displayText: "Quote Sections", linkRoute: "/framework", linkNumber: 211},
                    {displayText: "Brands", linkRoute: "/brands", linkNumber: 212},
                ]
            },
        ]
    },

];

export const colourStyles = [
    {background: "bg-col-01", colour: "col-01", border: "border-col-01"},
    {background: "bg-col-02", colour: "col-02", border: "border-col-02", transition: "transition-01-02"},
    {background: "bg-col-03", colour: "col-03", border: "border-col-03", transition: "transition-02-03"},
    {background: "bg-col-04", colour: "col-04", border: "border-col-04", transition: "transition-03-04"},
    {background: "bg-col-05", colour: "col-05", border: "border-col-05", transition: "transition-04-05"},
    {background: "bg-col-06", colour: "col-06", border: "border-col-06", transition: "transition-05-06"},
    {background: "bg-col-07", colour: "col-07", border: "border-col-07", transition: "transition-06-07"},
    {background: "bg-col-08", colour: "col-08", border: "border-col-08", transition: "transition-07-08"},
    {background: "bg-col-09", colour: "col-09", border: "border-col-09", transition: "transition-08-09"},
    {background: "bg-col-10", colour: "col-10", border: "border-col-10", transition: "transition-09-10"},
    {background: "bg-col-11", colour: "col-11", border: "border-col-11", transition: "transition-10-11"},
    {background: "bg-col-12", colour: "col-12", border: "border-col-12", transition: "transition-11-12"},
    {background: "bg-col-13", colour: "col-13", border: "border-col-13", transition: "transition-12-13"},
];