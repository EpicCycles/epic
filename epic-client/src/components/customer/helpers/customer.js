import {findIndexOfObjectWithId, generateRandomCode, updateObject} from "../../../helpers/utils";

export const updateCustomerBasic = (customer = {}, updatedValues) => {
    return updateObject(customer, updatedValues);
};

export const updateCustomerPhoneList = (customer = {}, phones) => {
    let updatedCustomer = updateObject(customer);
    updatedCustomer.phones = phones;
    if (customer.phones.length < phones.length) updatedCustomer.newPhone = {"dummyKey": generateRandomCode()};
    return updatedCustomer;
};

export const updateCustomerAddressList = (customer = {}, addresses) => {
    let updatedCustomer = updateObject(customer);
    updatedCustomer.addresses = addresses;
    if (customer.addresses && (customer.addresses.length < addresses.length)) updatedCustomer.newAddress = {"dummyKey": generateRandomCode()};
    return updatedCustomer;
};

export const customerAddErrorForAddress = (customer = {}, customerAddress) => {

    let updatedCustomer = updateObject(customer);

    const addressId = customerAddress.id;
    if (addressId && updatedCustomer.addresses) {
        const indexOfAddress = findIndexOfObjectWithId(updatedCustomer.addresses, addressId);
        updatedCustomer.addresses[indexOfAddress].error = true;
    } else {
        updatedCustomer.newAddress = customerAddress;
        updatedCustomer.newAddress.error = true;
    }
    return updatedCustomer;
};

export const customerAddErrorForPhone = (customer = {}, customerPhone) => {
    let updatedCustomer = updateObject(customer);

    const phoneId = customerPhone.id;
    if (phoneId && updatedCustomer.phones) {
        const indexOfPhone = findIndexOfObjectWithId(updatedCustomer.phones, phoneId);
        updatedCustomer.phones[indexOfPhone].error = true;
    } else {
        updatedCustomer.newPhone = customerPhone;
        updatedCustomer.newPhone.error = true;
    }
    return updatedCustomer;
};

