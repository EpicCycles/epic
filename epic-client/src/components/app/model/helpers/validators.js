import {INVALID_POSTCODE} from "./error";
import {POSTCODE_RULES} from "./constants";
export const validateData = (fieldList, currentValues) => {
    let errors = {};
    fieldList.forEach(field => {
        if (field.required) {
            if (!currentValues[field.fieldName]) {
                errors[field.fieldName] = field.error;
            }
        }
    });
    return errors;
};

export const validateEmailFormat = (email) => {
    const emailPattern = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return emailPattern.test(email);
};

export const validatePostcodeAndReturnError = (postcode, model = {}) => {
    let error;
    const postcodeRule = POSTCODE_RULES.filter(rule => rule.countryCode === model.country);
    if (postcodeRule[0]) {
        const OK = RegExp(postcodeRule[0].regex).exec(postcode);
        if (!OK) error = INVALID_POSTCODE + postcodeRule[0].display;
    }
    return error;
};