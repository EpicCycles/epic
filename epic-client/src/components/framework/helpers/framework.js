import {findObjectWithKey, updateObject} from "../../../helpers/utils";
import {NEW_ELEMENT_ID} from "../../../helpers/constants";
import {PART_TYPE_NAME_MISSING} from "../../app/model/helpers/error";

export const NEW_PART_TYPE = {
    attributes: [],
    name: "",
    can_be_substituted: true,
    can_be_omitted: true,
    customer_visible: true,
};
export const NEW_ATTRIBUTE = {
    attribute_name: "",
    in_use: true,
    mandatory: false,
    attribute_type: 1,
};
export const getPartTypeName = (partTypeId, sections) => {
    if (! partTypeId) return undefined;
    let partTypeName = "Unknown";
    sections.some(section => {
        return section.partTypes.some(partType => {
            if (partType.id === partTypeId) {
                partTypeName = partType.name;
                return true;
            }
            return false;
        });
    });
    return partTypeName;
};

export const processPartTypeValueChanges = (partType, componentKey, fieldName, input) => {
    const updatedPartType = updateObject(partType);
    if (fieldName.startsWith('name')) updatedPartType.name = input;
    if (!updatedPartType.name) {
        updatedPartType.error = true;
        updatedPartType.error_detail = PART_TYPE_NAME_MISSING;
    } else {
        updatedPartType.error = false;
        updatedPartType.error_detail = "";
    }
    if (fieldName.startsWith('description')) updatedPartType.description = input;
    if (fieldName.startsWith('can_be_substituted')) updatedPartType.can_be_substituted = input;
    if (fieldName.startsWith('can_be_omitted')) updatedPartType.can_be_omitted = input;
    if (fieldName.startsWith('customer_visible')) updatedPartType.customer_visible = input;
    if (fieldName.startsWith('attributes')) updatedPartType.attributes = input;
    if (fieldName.startsWith('synonyms')) updatedPartType.synonyms = input;
    if (fieldName.startsWith('detail')) updatedPartType._detail = input;
    if (componentKey === NEW_ELEMENT_ID) updatedPartType.dummyKey = NEW_ELEMENT_ID;

    updatedPartType.changed = true;
    return updatedPartType;
};
export const doesFieldMatchPartType = (partType, fieldName) => {
    const fieldNameLower = fieldName.toLowerCase();
    if (partType.name.toLowerCase() === fieldNameLower) {
        return true;
    } else {
        return partType.synonyms.some(synonym => {
            return (synonym.name.toLowerCase() === fieldNameLower);
        });
    }
};
export const attributeSummary = (attribute) => {
    let attributeDetail = [attribute.attribute_name];
    if (attribute.in_use) attributeDetail.push(" in use");
    if (attribute.mandatory) attributeDetail.push(" must be entered");
    if (attribute.options && attribute.options.length > 0) {
        attributeDetail.push(` allowed options: ${attribute.options.map((option) => {
            return option.option_name;
        }).toString()}`);
    }
    return attributeDetail.toString();
};
export const moveObjectUpOnePlace = (arrayOfObjects, objectId) => {
    const objectToMove = findObjectWithKey(arrayOfObjects, objectId);
    const currentIndex = arrayOfObjects.indexOf(objectToMove);
    if (currentIndex > 0) {
        let finalArray = arrayOfObjects;
        finalArray.splice(currentIndex, 1);
        const newPosition = currentIndex - 1;
        finalArray.splice(newPosition, 0, objectToMove);
        return resetPlacing(finalArray);
    } else {
        return arrayOfObjects;
    }
};
export const moveObjectDownOnePlace = (arrayOfObjects, objectId) => {
    let finalArrayOfObjects = arrayOfObjects;
    const objectToMove = findObjectWithKey(finalArrayOfObjects, objectId);
    const currentIndex = arrayOfObjects.indexOf(objectToMove);
    if (objectToMove && (currentIndex < (arrayOfObjects.length - 1))) {
        finalArrayOfObjects.splice(currentIndex, 1);
        const newPosition = currentIndex + 1;
        finalArrayOfObjects.splice(newPosition, 0, objectToMove);
        return resetPlacing(finalArrayOfObjects);
    } else {
        return finalArrayOfObjects;
    }
};
export const moveObjectToTop = (arrayOfObjects, objectId) => {
    let finalArrayOfObjects = arrayOfObjects;
    const objectToMove = findObjectWithKey(finalArrayOfObjects, objectId);
    const currentIndex = arrayOfObjects.indexOf(objectToMove);
    if (currentIndex > 0) {
        finalArrayOfObjects.splice(currentIndex, 1);
        finalArrayOfObjects.unshift(objectToMove);
        return resetPlacing(finalArrayOfObjects);
    } else {
        return finalArrayOfObjects;
    }
};
export const moveObjectToBottom = (arrayOfObjects, objectId) => {
    let finalArrayOfObjects = arrayOfObjects;
    const objectToMove = findObjectWithKey(finalArrayOfObjects, objectId);
    const currentIndex = arrayOfObjects.indexOf(objectToMove);
    if (objectToMove && (currentIndex < (arrayOfObjects.length - 1))) {
        finalArrayOfObjects.splice(currentIndex, 1);
        finalArrayOfObjects.push(objectToMove);
        return resetPlacing(finalArrayOfObjects);
    } else {
        return finalArrayOfObjects;
    }
};
export const renumberAll = (sections) => {
    let sectionSequence = 10;
    return sections.map((section) => {
        section.placing = sectionSequence;
        sectionSequence = sectionSequence + 10;
        let partTypeSequence = 10;
        section.partTypes = section.partTypes ? section.partTypes.map((partType) => {
            partType.placing = partTypeSequence;
            partTypeSequence = partTypeSequence + 10;
            let attributeSequence = 10;
            partType.attributes = partType.attributes ? partType.attributes.map((attribute) => {
                attribute.placing = attributeSequence;
                attribute.options = attribute.options ? resetPlacing(attribute.options) : [];
                attributeSequence = attributeSequence + 10;
                return attribute;
            }) : [];
            return partType;
        }) : section.partTypes = [];

        return section;
    })
};

export const resetPlacing = (arrayOfPlacedObjects) => {
    let objectPlacing = 10;
    return arrayOfPlacedObjects.map((placedObject) => {
        placedObject.placing = objectPlacing;
        objectPlacing = objectPlacing + 10;
        return placedObject;
    })
};

