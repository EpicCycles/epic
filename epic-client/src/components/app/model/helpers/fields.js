import {
    ADDRESS_MISSING,
    BRAND_MISSING, BRAND_NAME_MISSING,
    BUNDLE_NAME_MISSING, COUNTRY_MISSING,
    FRAME_NAME_MISSING,
    MODEL_NAME_MISSING,
    PART_MISSING,
    PART_NAME_MISSING,
    PART_TYPE_MISSING,
    PRODUCTS_MISSING,
    SUPPLIER_MISSING
} from "./error";
import {validatePostcodeAndReturnError} from "./validators";

// TODO customer fields including email with email validator
export const ADDRESS1 = "address1";
export const ADDRESS2 = "address2";
export const ADDRESS3 = "address3";
export const ADDRESS4 = "address4";
export const COUNTRY = "country";
export const POSTCODE = "postcode";
export const BRAND = "brand";
export const SUPPLIER = "supplier";
export const PART = "part";
export const LINK = "link";
export const BUNDLE = "bundle_name";
export const PRODUCT_CODE = "product_code";
export const PART_TYPE = "partType";
export const BRAND_NAME = "brand_name";
export const BIKE_BRAND = "bike_brand";
export const FRAME_NAME = "frame_name";
export const MODEL_NAME = "model_name";
export const PART_NAME = "part_name";
export const DESCRIPTION = "description";
export const COLOURS = "colours";
export const SELL_PRICE = "rrp";
export const EPIC_PRICE = "epic_price";
export const TICKET_PRICE = "ticket_price";
export const TRADE_PRICE = "trade_price";
export const CLUB_PRICE = "club_price";
export const FITTED_PRICE = "fitted_price";
export const TRADE_IN_PRICE = "trade_in_value";
export const SIZES = "sizes";
export const STANDARD = "standard";
export const STOCKED = "stocked";
export const PRODUCTS = "products";
export const TEXT = "text";
export const NUMBER = "number";
export const CURRENCY = "currency";
export const TEXT_AREA = "textArea";
export const CHECKBOX = "checkbox";

// TODO add COUNTRY type and fields for address
// TODO consider having a function passed for validator
export const ADDRESS1_FIELD = {
    fieldName: ADDRESS1,
    type: TEXT,
    length: 100,
    header: "Address",
    required: true,
    error: ADDRESS_MISSING
};
export const ADDRESS2_FIELD = {
    fieldName: ADDRESS2,
    type: TEXT,
    length: 100,
    header: "Line 2",
};
export const ADDRESS3_FIELD = {
    fieldName: ADDRESS3,
    type: TEXT,
    length: 100,
    header: "Line 3",
};
export const ADDRESS4_FIELD = {
    fieldName: ADDRESS4,
    type: TEXT,
    length: 100,
    header: "Line 4",
};
export const COUNTRY_FIELD = {
    fieldName: COUNTRY,
    header: "Country",
    type: COUNTRY,
    required: true,
    error: COUNTRY_MISSING
};
export const POSTCODE_FIELD = {
    fieldName: POSTCODE,
    type: TEXT,
    length: 20,
    header: "Postcode",
    validator: validatePostcodeAndReturnError,
    validatorAdditionalFields: [COUNTRY]
};
export const BRAND_FIELD = {
    fieldName: BRAND,
    header: "Brand",
    synonyms: [BRAND, 'manufacturer'],
    required: true,
    error: BRAND_MISSING,
    type: BRAND
};
export const PRODUCTS_FIELD = {
    fieldName: PRODUCTS,
    header: "Products",
    synonyms: [],
    required: true,
    error: PRODUCTS_MISSING,
    type: PRODUCTS
};
export const SUPPLIER_FIELD = {
    fieldName: SUPPLIER,
    header: "Supplier",
    synonyms: [SUPPLIER],
    required: true,
    error: SUPPLIER_MISSING,
    type: SUPPLIER
};
export const SUPPLIER_FIELD_OPTIONAL = {
    fieldName: SUPPLIER,
    header: "Supplier",
    synonyms: [SUPPLIER],
    type: SUPPLIER
};
export const PART_FIELD = {
    fieldName: PART,
    header: "Part",
    synonyms: [PART],
    required: true,
    error: PART_MISSING,
    type: PART
};
export const FRAME_NAME_FIELD = {
    fieldName: FRAME_NAME,
    header: "Frame Name",
    synonyms: [],
    required: true,
    error: FRAME_NAME_MISSING
};
export const LINK_FIELD = {
    fieldName: LINK,
    header: "URL",
    synonyms: [],
};
export const BRAND_NAME_FIELD = {
    fieldName: BRAND_NAME,
    header: "Brand Name",
    synonyms: [],
    required: true,
    error: BRAND_NAME_MISSING
};
export const BUNDLE_NAME_FIELD = {
    fieldName: BUNDLE,
    header: "Bundle Name",
    synonyms: [],
    required: true,
    error: BUNDLE_NAME_MISSING
};
export const MODEL_NAME_FIELD = {
    fieldName: MODEL_NAME,
    header: "Model Name",
    synonyms: [],
    required: true,
    error: MODEL_NAME_MISSING,
    type: TEXT,
    length: 100
};
export const COLOURS_FIELD = {
    fieldName: COLOURS,
    header: "Colours",
    synonyms: [COLOURS, "colour", "colors", "color"],
    type: TEXT,
    length: 100
};
export const DESCRIPTION_FIELD = {
    fieldName: DESCRIPTION,
    header: "Description",
    synonyms: [DESCRIPTION, "desc"],
    type: TEXT_AREA,
    length: 400
};
export const PRODUCT_CODE_FIELD = {
    fieldName: PRODUCT_CODE,
    header: "Product Code",
    synonyms: [PRODUCT_CODE],
    type: TEXT,
    length: 30
};
export const SELL_PRICE_FIELD = {
    fieldName: SELL_PRICE,
    header: "RRP",
    synonyms: ["price", "selling price", "srp", "rrp", "sell price", "retail price"],
    type: CURRENCY,
    length: 10
};
export const EPIC_PRICE_FIELD = {
    fieldName: EPIC_PRICE,
    header: "Epic Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const CLUB_PRICE_FIELD = {
    fieldName: CLUB_PRICE,
    header: "Club Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};

export const FITTED_PRICE_FIELD = {
    fieldName: FITTED_PRICE,
    header: "Fitted Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const TRADE_PRICE_FIELD = {
    fieldName: TRADE_PRICE,
    header: "Trade Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const TICKET_PRICE_FIELD = {
    fieldName: TICKET_PRICE,
    header: "Ticket Price",
    synonyms: [],
    type: CURRENCY,
    length: 10
};
export const SIZES_FIELD = {
    fieldName: SIZES,
    header: "Sizes",
    synonyms: [SIZES, "size", "frame sizes", "frame size"],
    type: TEXT,
    length: 100
};
export const PART_TYPE_FIELD = {
    fieldName: PART_TYPE,
    header: "Part Type",
    required: true,
    error: PART_TYPE_MISSING,
    type: PART_TYPE,
};
export const PART_NAME_FIELD = {
    fieldName: PART_NAME,
    header: "Part Name",
    required: true,
    error: PART_NAME_MISSING
};
export const TRADE_IN_FIELD = {
    fieldName: TRADE_IN_PRICE,
    header: "Trade In Â£",
    type: CURRENCY,
    length: 10
};
export const STOCKED_FIELD = {
    fieldName: STOCKED,
    header: "Stocked",
    type: CHECKBOX
};
export const STANDARD_FIELD = {
    fieldName: STANDARD,
    header: "Standard",
    type: CHECKBOX
};
export const BIKE_BRAND_FIELD = {
    fieldName: BIKE_BRAND,
    header: "Bike Brand",
    type: CHECKBOX
};
export const customerAddressFields = [ADDRESS1_FIELD, ADDRESS2_FIELD, ADDRESS3_FIELD, ADDRESS4_FIELD, COUNTRY_FIELD, POSTCODE_FIELD];
export const frameFields = [BRAND, FRAME_NAME];
export const brandFields = [BRAND_NAME_FIELD, BIKE_BRAND_FIELD, LINK_FIELD, SUPPLIER_FIELD_OPTIONAL];
export const bikeFields = [MODEL_NAME_FIELD, DESCRIPTION_FIELD, COLOURS_FIELD, SELL_PRICE_FIELD, EPIC_PRICE_FIELD, CLUB_PRICE_FIELD, SIZES_FIELD];
export const partFields = [
    PART_TYPE_FIELD,
    BRAND_FIELD,
    PART_NAME_FIELD,
    TRADE_IN_FIELD,
    STANDARD_FIELD,
    STOCKED_FIELD
];
export const supplierProductFields = [
    SUPPLIER_FIELD,
    FITTED_PRICE_FIELD,
    TICKET_PRICE_FIELD,
    SELL_PRICE_FIELD,
    TRADE_PRICE_FIELD,
    CLUB_PRICE_FIELD
];
export const bundleFields = [
    BUNDLE_NAME_FIELD,
    PRODUCTS_FIELD,
    FITTED_PRICE_FIELD,
    TICKET_PRICE_FIELD,
];

