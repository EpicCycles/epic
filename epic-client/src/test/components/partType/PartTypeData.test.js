import React from 'react';
import PartTypeData from "../../../components/partType/PartTypeData";
import {NEW_ELEMENT_ID} from "../../../helpers/constants";
test("displays the PartTypeData component for an existing part type with attributes, detail hidden", () => {
    const componentKey = "21";
    const partType = {
        _detail: false,
        id:21,
        shortName:"part one",
        can_be_substituted: false,
        can_be_omitted: true,
        customer_facing:true
    };
    const component = shallow(<PartTypeData componentKey={componentKey} partType={partType} />);
    expect(component).toMatchSnapshot();
});
test("displays the PartTypeData component for an existing part type with attributes, detail shown", () => {
    const componentKey = "dummy2";
    const partType = {
        _detail: true,
        dummyKey:componentKey,
        shortName:"part one",
        can_be_substituted: false,
        can_be_omitted: true,
        customer_facing:true,
        attributes: ["attribute1","attribute2"],
        synonyms: ["synonym1","synonym1"]
    };
    const component = shallow(<PartTypeData componentKey={componentKey} partType={partType} />);
    expect(component).toMatchSnapshot();
});
test("displays the PartTypeData component for a new part type with no attributes", () => {
    const partType = {
        _detail: true,
        dummyKey:NEW_ELEMENT_ID,
        shortName:"part one",
        can_be_substituted: false,
        can_be_omitted: true,
        customer_facing:true
};
    const component = shallow(<PartTypeData componentKey={NEW_ELEMENT_ID} partType={partType} />);
    expect(component).toMatchSnapshot();
});