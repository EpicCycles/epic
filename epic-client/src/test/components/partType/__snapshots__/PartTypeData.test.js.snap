// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays the PartTypeData component for a new part type with no attributes 1`] = `
<React.Fragment>
  <FormTextInput
    fieldName="shortName_new"
    onClick={[Function]}
    placeholder="add new"
    value="part one"
  />
  <label
    htmlFor="can_be_substituted_new"
  >
    Can Be Substituted?
  </label>
  <input
    checked={false}
    id="can_be_substituted_new"
    name="can_be_substituted_new"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="can_be_omitted_new"
  >
    Can Be Omitted?
  </label>
  <input
    checked={true}
    id="can_be_omitted_new"
    name="can_be_omitted_new"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="customer_facing_new"
  >
    Customer Facing?
  </label>
  <input
    checked={true}
    id="customer_facing_new"
    name="customer_facing_new"
    onChange={[Function]}
    type="checkbox"
  />
  Attributes: 0
  Synonyms: 0
  <React.Fragment>
    <PartTypeAttributes
      attributes={Array []}
      partTypeKey="new"
    />
    <PartTypeSynonyms
      partTypeKey="new"
      synonyms={Array []}
    />
  </React.Fragment>
</React.Fragment>
`;

exports[`displays the PartTypeData component for an existing part type with attributes, detail hidden 1`] = `
<React.Fragment>
  <FormTextInput
    fieldName="shortName_21"
    onClick={[Function]}
    placeholder="add new"
    value="part one"
  />
  <label
    htmlFor="can_be_substituted_21"
  >
    Can Be Substituted?
  </label>
  <input
    checked={false}
    id="can_be_substituted_21"
    name="can_be_substituted_21"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="can_be_omitted_21"
  >
    Can Be Omitted?
  </label>
  <input
    checked={true}
    id="can_be_omitted_21"
    name="can_be_omitted_21"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="customer_facing_21"
  >
    Customer Facing?
  </label>
  <input
    checked={true}
    id="customer_facing_21"
    name="customer_facing_21"
    onChange={[Function]}
    type="checkbox"
  />
  Attributes: 0
  Synonyms: 0
</React.Fragment>
`;

exports[`displays the PartTypeData component for an existing part type with attributes, detail shown 1`] = `
<React.Fragment>
  <FormTextInput
    fieldName="shortName_dummy2"
    onClick={[Function]}
    placeholder="add new"
    value="part one"
  />
  <label
    htmlFor="can_be_substituted_dummy2"
  >
    Can Be Substituted?
  </label>
  <input
    checked={false}
    id="can_be_substituted_dummy2"
    name="can_be_substituted_dummy2"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="can_be_omitted_dummy2"
  >
    Can Be Omitted?
  </label>
  <input
    checked={true}
    id="can_be_omitted_dummy2"
    name="can_be_omitted_dummy2"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="customer_facing_dummy2"
  >
    Customer Facing?
  </label>
  <input
    checked={true}
    id="customer_facing_dummy2"
    name="customer_facing_dummy2"
    onChange={[Function]}
    type="checkbox"
  />
  Attributes: 2
  Synonyms: 2
  <React.Fragment>
    <PartTypeAttributes
      attributes={
        Array [
          "attribute1",
          "attribute2",
        ]
      }
      partTypeKey="dummy2"
    />
    <PartTypeSynonyms
      partTypeKey="dummy2"
      synonyms={
        Array [
          "synonym1",
          "synonym1",
        ]
      }
    />
  </React.Fragment>
</React.Fragment>
`;
